### ISSUE TITLE
Push branch without checking out.

### ISSUE THREAD

**Issue by uloco** [jesseduffield/lazygit#2755]
I'd like to be able to push a branch without checking out, by just highlighting the branch and pressing `P`.

Especially useful when you have updateRefs config enabled and rebase a bunch of stacked branches and need to push all of them afterwards. But in general I think it is better UX to make actions in the branch panel concern the branch I am highlighting. It  works for fetch and update / rebase / merge like this. Why not push?

**Comment by uloco** [comment]
two very true and important points, agree!

**Comment by stefanhaller** [comment]
For the first one, a possible mitigation could be to introduce a confirmation prompt only for the case that your are in the branches panel. We had a discussion recently whether `P` should have a confirmation prompt, and I voted against it, but for this particular case it could be justified. (On the other hand, it could also be confusing that the same keyboard command sometimes prompts and other times doesn't, depending on where you are.)

**Comment by rukai** [comment]
A possible solution I thought of that avoids dangerous changes to the interface:
If a different branch to the checked out branch is selected, and a push is attempted, lazygit should fail the push with the error "Cannot push a branch that is not checked out".
This will ensure the user realizes they are pushing the checked out branch and not the selected branch.

**Comment by jesseduffield** [comment]
@rukai isn't what you're describing the use case of this issue? I.e. selecting a branch that is not checked out and trying to push it?

@stefanhaller regarding your second point above, we could have a stopgap approach of only allowing the push if the branch is already tracking a remote.

Another thought that just came to mind: perhaps we keep P/p for simply pushing/pulling for the checked out branch, and have a new 'sync' menu (keybinding: 's') openable from the branches view which lets you do more specific versions of fetch/push/pull. This could be a place for all  the complicated edge-casey things we sometimes need to do with these operations, which would cost us an extra key if supported through the global P/p keys.

The caveat is that we already have a precedent of a key working differently depending on where you press it: 'f' will fetch for the current branch if you're in the files view, or the selected branch if you're in the branches view. That doesn't seem to trip people up, but it's also a safe operation.

Whaddayathink?

**Comment by stefanhaller** [comment]
> @stefanhaller regarding your second point above, we could have a stopgap approach of only allowing the push if the branch is already tracking a remote.

That could help a bit, but maybe not enough. One common workflow that I find myself doing quite often is to rebase the current branch onto master and then force-push. To do that, I have my branch checked out, go to the branches panel, select master and hit "r", and then shift-P to push. I'm sure I could get used to having to select my branch again first (or go to another panel), so no huge objection if we want to go that route.

> Another thought that just came to mind: perhaps we keep P/p for simply pushing/pulling for the checked out branch, and have a new 'sync' menu (keybinding: 's') openable from the branches view which lets you do more specific versions of fetch/push/pull. 

I like the idea. One operation that I would want to have in such a menu would be to pull (not just fetch) the selected branch even if it has diverged from upstream. It's a common scenario to pull all branches of a stack that has been rebased and force-pushed by a colleague, and it's annoying to have to check out each branch first. Not sure how this would be implemented; maybe it would have to check out the branch behind the scenes, pull it, and switch back to the previous branch.

But wait, if we can do that, then we might as well put it on "p" in the branches panel, so that's not really an argument for the sync menu.

> The caveat is that we already have a precedent of a key working differently depending on where you press it: 'f' will fetch for the current branch if you're in the files view, or the selected branch if you're in the branches view. That doesn't seem to trip people up, but it's also a safe operation.

Good point. I use both all the time actually, and I have never been confused about them doing unexpected things when I'm in the wrong panel, so I suppose I could get used to this for shift-P too.

**Comment by jesseduffield** [comment]
> Not sure how this would be implemented; maybe it would have to check out the branch behind the scenes, pull it, and switch back to the previous branch.

I think our soon-to-be-added worktree support can help here. We could have a hidden worktree created in some temp directory (out of the user's way) that can be used for checking out a branch, doing something, and then detaching again to be reused down the line. This would be much more reliable than trying to autostash changes on the current worktree, check out the branch, etc. The issue is that some repos are huge and users may not want a copy of the repo sitting on their computer for occasional use by lazygit, no would they want to wait for a transient worktree to be created and removed each time they want to pull a branch. But a work-tree approach really would open up a lot of possibilities.

**Comment by stefanhaller** [comment]
Interesting thought. If it's a hidden worktree, I wonder why we need to wait for lazygit's worktree support though. Can't we just do what you suggest behind the scenes, now? Or is it for the case that the pull fails with a conflict, and we want to expose the worktree to the user to sort it out themselves? My instinct would have been to abort the whole thing in that case, show an error, and let the user checkout the branch manually to pull it.

Another thing I'm not sure about is if the pull would even work in a different worktree. If a branch diverged from upstream, then git relies on the reflog in some magic way to figure out how exactly to rebase it to avoid conflicts, resulting in essentially a hard reset to upstream. Not sure if reflogs are shared between worktrees. But I also have almost zero experience with worktrees, never found a reason to use them...

**Comment by jesseduffield** [comment]
Good point: we can do this right now.

From what I've seen, worktrees indeed have their own reflogs, though I wasn't aware of git pull using the reflog, now can I see anything about it online.

**Comment by stefanhaller** [comment]
> though I wasn't aware of git pull using the reflog, now can I see anything about it online.

I tried to find a reference for this, and couldn't find one (not even in the git source code). I must have dreamed this, sorry for the confusion. There is some smartness that git uses to better rebase the branch in case the upstream branch was rebased since the last time it was fetched, but it doesn't use the reflog.

**Comment by jesseduffield** [comment]
No need to apologise, good to know :)

**Comment by stefanhaller** [comment]
> We could have a hidden worktree created in some temp directory (out of the user's way) that can be used for checking out a branch, doing something, and then detaching again to be reused down the line.

I looked into this a little bit, but I don't see how this would work (the reuse bit). As far as I can see, there's no way to detach/reattach worktrees to reuse them. So we'd have to have our hidden worktree active all the time; maybe we could hide it in whatever UI we're going to add for worktrees, but it would still be visible if you do `git worktree list` on the command line, which is not ideal. (Although if we give it a clear enough name, it might be acceptable.)

Or am I missing anything here? I'm not experienced with worktrees.

**Comment by jesseduffield** [comment]
The connection between a git repo and one of it's linked worktrees is established by the presence of at `.git/worktrees/<worktreename>`. So when we want to detach the worktree we would just move that folder to somewhere managed by lazygit.

**Comment by jesseduffield** [comment]
(For the record this might end up being an overcomplicated solution to the problem but it's worth exploring)

**Comment by stefanhaller** [comment]
> The connection between a git repo and one of it's linked worktrees is established by the presence of at `.git/worktrees/<worktreename>`. So when we want to detach the worktree we would just move that folder to somewhere managed by lazygit.

I would feel more comfortable doing these things through git commands rather than messing with git's internal state, and there doesn't seem to be a way to do that through commands.

**Comment by jesseduffield** [comment]
Fair point: I should say that as part of the worktrees PR I'm already messing with its internal state (reading it at least) to overcome shortcomings in the commands

**Comment by jesseduffield** [comment]
I think if we're gonna do a POC for this we should start with one that just creates and removes a worktree rather than trying to re-use one, because if you're pushing as part of dealing with stacked branches, it's not very time critical

**Comment by stefanhaller** [comment]
I disagree; it would be quite annoying to have to wait longer than necessary for the branches to get pushed (or pulled). And creating and removing a worktree every time would be totally infeasible for our monorepo at work, it would just take too long.

**Comment by jesseduffield** [comment]
Snap :( So our choices are:
* stash/unstash and perform the action in the current worktree
 - can't do many pushes at once
 - annoying if you don't want your changes stashed or if something breaks halfway through
 - same behaviour as regular push
* re-use a worktree
  - can't do many pushes at once
  - user's current worktree not affected if something goes wrong
  - requires messing with internal git files to achieve without showing the worktree
   - same behaviour as regular push (pretty sure)
* create/destroy worktree as needed
  - can do many pushes at once
  - can be very slow
  - same behaviour as regular push (pretty sure)  
* regular git-push command
  - fast
  - fault-tolerant
  - not necessarily the same behaviour as pushing current branch
  
Each of these options has pros/cons. I'm leaning towards the last one.

**Comment by stefanhaller** [comment]
For pushing I agree, I'd go with the last one for that, too. And I think it would be possible to emulate what git does for the current branch to make the behavior very similar.

However, the worktree discussion started because we also want to be able to _pull_ a branch that's not checked out, and there doesn't seem to be a similar approach for that.

**Comment by stefanhaller** [comment]
I still think it's interesting to explore the second option (reusing worktrees), and if we can make it work for pulling, then use it for pushing as well, just to avoid having to emulate git's push-current-branch behavior.

BTW, pulling/pushing multiple branches at once is not really a requirement for me (and it would be unclear what the UX for that would be exactly). It would be a big improvement already if I can just pull/push each branch in the stack one by one, as long as I don't have to check them out.

**Comment by jesseduffield** [comment]
Alright sounds good

**Comment by NGuldager** [comment]
I was about to add a new feature request. But I thought it might be a good idea to make a quick search if there was something similar already asked for. I think this one is close enough that i'll just add mine as a comment here.

### Pushing multiple branches at once

**Is your feature request related to a problem? Please describe.**
I am constantly working on stacked branches. And because of this I often need to push multiple branches when I make updates based on pull request comments.

**Describe the solution you'd like**
The ideal solution would be to just have built in support for pushing multiple branches in one go. This could be by using the already built in selection ("visual" mode or range select) of multiple branches and allowing a push from there to push multiple branches. 

**Describe alternatives you've considered**
I started looking into a custom command, but I can only get 1 branch name from the `SelectedLocalBranch`. Another solution could be to allow the user to get the full list of all branches selected for custom commands, then it would be fairly simple create a custom command for pushing multiple branches

**Comment by stefanhaller** [comment]
I finally got around to improving the situation of pushing a stack of branches; see #6045. The way this works is that you check out the tip of the stack (where you normally work anyway if you use branch stacks), push it normally using shift-P, and it will ask you if you also want to push the other branches of the stack. I found this approach more promising than adding a command to push the selected branch(es). Feedback welcome.

This only addresses pushing; the opposite (pulling a stack that a co-worker has rebased) is still unsolved. I have ideas about that too, but I don't need that quite as often, so it has lower priority for me.
